Orange/widgets/data/tests/test_owdatasampler.py

```diff
@@ -26,3 +26,36 @@ def test_error_message(self):
         self.assertFalse(self.widget.Error.too_many_folds.is_shown())
         self.send_signal("Data", Table(self.iris.domain))
         self.assertTrue(self.widget.Error.no_data.is_shown())
+
+    def test_stratified_on_unbalanced_data(self):
+        unbalanced_data = self.iris[:51]
+
+        self.widget.controls.stratify.setChecked(True)
+        self.send_signal("Data", unbalanced_data)
+        self.assertTrue(self.widget.Warning.could_not_stratify.is_shown())
+
+    def test_bootstrap(self):
+        self.select_sampling_type(self.widget.Bootstrap)
+
+        self.send_signal("Data", self.iris)
+
+        in_input = set(self.iris.ids)
+        sample = self.get_output("Data Sample")
+        in_sample = set(sample.ids)
+        in_remaining = set(self.get_output("Remaining Data").ids)
+
+        # Bootstrap should sample len(input) instances
+        self.assertEqual(len(sample), len(self.iris))
+        # Sample and remaining should cover all instances, while none
+        # should be present in both
+        self.assertEqual(len(in_sample | in_remaining), len(in_input))
+        self.assertEqual(len(in_sample & in_remaining), 0)
+        # Sampling with replacement will always produce at least one distinct
+        # instance in sample, and at least one instance in remaining with
+        # high probability (1-(1/150*2/150*...*145/150) ~= 1-2e-64)
```

Comment:
But just in case we ever need it:
```python
def assertProbablyGreater(self, a, b, msg=None):
    """Assert that a in greater than b in at least one out of two runs."""
    import inspect

    try:
        return self.assertGreater(a, b, msg)
    except AssertionError:
        pass

    frames = inspect.stack()[1:]
    test_frames = [f for f in frames if f.function.startswith("test_")]
    if any(f for f in frames if f.function == "c") or not test_frames:
        # Retried, but still not working or nothing to retry. Fail
        return self.assertGreater(a, b, msg)

    getattr(self, test_frames[0].function)()
```